File: client-sdk/src/services/risk.ts
/**
 * Risk Assessment Service
 * Wraps /risk endpoints for transaction risk scoring
 */

import { AxiosInstance } from 'axios';
import { EventEmitter } from '../events';
import { BaseService } from './base';

// Local type definitions to avoid circular imports
export type RiskLevel = 'low' | 'medium' | 'high' | 'critical';

export interface LabelInfo {
  label: string;
  category: string;
  severity: string;
  source: string;
}

export interface RiskScore {
  address: string;
  score: number;
  level: RiskLevel;
  timestamp: string;
  expiresAt: string;
}

export interface RiskAssessmentRequest {
  address: string;
  transactionHash?: string;
  amount?: string;
  counterparty?: string;
  metadata?: Record<string, any>;
}

export interface RiskAssessmentResponse {
  address: string;
  riskScore: number;
  riskLevel: RiskLevel;
  factors: RiskFactor[];
  labels: LabelInfo[];
  timestamp: string;
  expiresAt: string;
  cached: boolean;
}

export interface RiskFactor {
  name: string;
  weight: number;
  value: number;
  description: string;
}

export interface BatchRiskRequest {
  addresses: string[];
  includeLabels?: boolean;
  includeFactors?: boolean;
}

export interface BatchRiskResponse {
  results: RiskAssessmentResponse[];
  processedCount: number;
  failedCount: number;
  failures: { address: string; error: string }[];
}

export interface RiskThreshold {
  level: RiskLevel;
  minScore: number;
  maxScore: number;
  action: 'allow' | 'review' | 'block';
}

export class RiskService extends BaseService {
  constructor(http: AxiosInstance, events: EventEmitter) {
    super(http, events);
  }

  /**
   * Assess risk for a single address
   */
  async assess(request: RiskAssessmentRequest): Promise<RiskAssessmentResponse> {
    const response = await this.http.post<RiskAssessmentResponse>('/risk/assess', request);
    
    this.events.emit('risk:assessed', {
      address: request.address,
      riskLevel: response.data.riskLevel,
      riskScore: response.data.riskScore
    });

    return response.data;
  }

  /**
   * Get cached risk score for an address
   */
  async getScore(address: string): Promise<RiskScore | null> {
    try {
      const response = await this.http.get<RiskScore>(`/risk/score/${address}`);
      return response.data;
    } catch (error: any) {
      if (error.response?.status === 404) {
        return null;
      }
      throw error;
    }
  }

  /**
   * Batch assess multiple addresses
   */
  async batchAssess(request: BatchRiskRequest): Promise<BatchRiskResponse> {
    const response = await this.http.post<BatchRiskResponse>('/risk/batch', request);
    
    this.events.emit('risk:batchCompleted', {
      processedCount: response.data.processedCount,
      failedCount: response.data.failedCount
    });

    return response.data;
  }

  /**
   * Get risk thresholds configuration
   */
  async getThresholds(): Promise<RiskThreshold[]> {
    const response = await this.http.get<{ thresholds: RiskThreshold[] }>('/risk/thresholds');
    return response.data.thresholds;
  }

  /**
   * Check if address passes risk threshold
   */
  async checkThreshold(address: string, maxRiskLevel: RiskLevel = 'medium'): Promise<{
    passed: boolean;
    riskScore: number;
    riskLevel: RiskLevel;
    action: 'allow' | 'review' | 'block';
  }> {
    const response = await this.http.post<{
      passed: boolean;
      riskScore: number;
      riskLevel: RiskLevel;
      action: 'allow' | 'review' | 'block';
    }>('/risk/check-threshold', { address, maxRiskLevel });
    
    return response.data;
  }

  /**
   * Get risk history for an address
   */
  async getHistory(address: string, options?: {
    limit?: number;
    offset?: number;
    startDate?: Date;
    endDate?: Date;
  }): Promise<{
    history: RiskAssessmentResponse[];
    total: number;
  }> {
    const params = new URLSearchParams();
    if (options?.limit) params.append('limit', options.limit.toString());
    if (options?.offset) params.append('offset', options.offset.toString());
    if (options?.startDate) params.append('startDate', options.startDate.toISOString());
    if (options?.endDate) params.append('endDate', options.endDate.toISOString());

    const response = await this.http.get<{
      history: RiskAssessmentResponse[];
      total: number;
    }>(`/risk/history/${address}?${params.toString()}`);

    return response.data;
  }

  /**
   * Invalidate cached risk score
   */
  async invalidateCache(address: string): Promise<void> {
    await this.http.delete(`/risk/cache/${address}`);
    this.events.emit('risk:cacheInvalidated', { address });
  }

  /**
   * Get risk factors configuration
   */
  async getFactors(): Promise<{
    factors: {
      name: string;
      weight: number;
      description: string;
      enabled: boolean;
    }[];
  }> {
    const response = await this.http.get('/risk/factors');
    return response.data;
  }
}

File: client-sdk-python/amttp/services/risk.py
"""Risk Assessment Service — wraps ``/risk`` endpoints."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel

from amttp.services.base import BaseService

# ── types ─────────────────────────────────────────────────────────────────────

RiskLevel = Literal["low", "medium", "high", "critical"]


class RiskFactor(BaseModel):
    name: str
    weight: float
    value: float
    description: str


class LabelInfo(BaseModel):
    label: str
    category: str
    severity: str
    source: str


class RiskAssessmentRequest(BaseModel):
    address: str
    transaction_hash: Optional[str] = None
    amount: Optional[str] = None
    counterparty: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


class RiskAssessmentResponse(BaseModel):
    address: str
    risk_score: float = 0
    risk_level: str = "low"
    factors: List[RiskFactor] = []
    labels: List[LabelInfo] = []
    timestamp: str = ""
    expires_at: str = ""
    cached: bool = False

    model_config = {"populate_by_name": True}


class RiskScore(BaseModel):
    address: str
    score: float
    level: str
    timestamp: str
    expires_at: str


class BatchRiskRequest(BaseModel):
    addresses: List[str]
    include_labels: Optional[bool] = None
    include_factors: Optional[bool] = None


class BatchRiskResponse(BaseModel):
    results: List[RiskAssessmentResponse] = []
    processed_count: int = 0
    failed_count: int = 0
    failures: List[Dict[str, str]] = []


class RiskThreshold(BaseModel):
    level: str
    min_score: float
    max_score: float
    action: Literal["allow", "review", "block"]


# ── service ───────────────────────────────────────────────────────────────────


class RiskService(BaseService):
    """Risk assessment operations."""

    async def assess(self, request: RiskAssessmentRequest) -> RiskAssessmentResponse:
        """Assess risk for a single address."""
        resp = await self._http.post("/risk/assess", json=request.model_dump(exclude_none=True))
        resp.raise_for_status()
        data = RiskAssessmentResponse.model_validate(resp.json())
        self._events.emit("risk:assessed", {
            "address": request.address,
            "risk_level": data.risk_level,
            "risk_score": data.risk_score,
        })
        return data

    async def get_score(self, address: str) -> Optional[RiskScore]:
        """Get cached risk score for an address."""
        from amttp.errors import AMTTPError, AMTTPErrorCode

        try:
            resp = await self._http.get(f"/risk/score/{address}")
            resp.raise_for_status()
            return RiskScore.model_validate(resp.json())
        except AMTTPError as e:
            if e.code == AMTTPErrorCode.NOT_FOUND:
                return None
            raise

    async def batch_assess(self, request: BatchRiskRequest) -> BatchRiskResponse:
        """Batch-assess multiple addresses."""
        resp = await self._http.post("/risk/batch", json=request.model_dump(exclude_none=True))
        resp.raise_for_status()
        data = BatchRiskResponse.model_validate(resp.json())
        self._events.emit("risk:batchCompleted", {
            "processed_count": data.processed_count,
            "failed_count": data.failed_count,
        })
        return data

    async def get_thresholds(self) -> List[RiskThreshold]:
        """Get risk threshold configuration."""
        resp = await self._http.get("/risk/thresholds")
        resp.raise_for_status()
        return [RiskThreshold.model_validate(t) for t in resp.json()["thresholds"]]

    async def check_threshold(
        self, address: str, max_risk_level: RiskLevel = "medium"
    ) -> Dict[str, Any]:
        """Check if address passes risk threshold."""
        resp = await self._http.post(
            "/risk/check-threshold",
            json={"address": address, "maxRiskLevel": max_risk_level},
        )
        resp.raise_for_status()
        return resp.json()

    async def get_history(
        self,
        address: str,
        *,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
    ) -> Dict[str, Any]:
        """Get risk history for an address."""
        params: Dict[str, str] = {}
        if limit is not None:
            params["limit"] = str(limit)
        if offset is not None:
            params["offset"] = str(offset)
        if start_date:
            params["startDate"] = start_date.isoformat()
        if end_date:
            params["endDate"] = end_date.isoformat()
        resp = await self._http.get(f"/risk/history/{address}", params=params)
        resp.raise_for_status()
        return resp.json()

    async def invalidate_cache(self, address: str) -> None:
        """Invalidate cached risk score."""
        resp = await self._http.delete(f"/risk/cache/{address}")
        resp.raise_for_status()
        self._events.emit("risk:cacheInvalidated", {"address": address})

    async def get_factors(self) -> Dict[str, Any]:
        """Get risk factors configuration."""
        resp = await self._http.get("/risk/factors")
        resp.raise_for_status()
        return resp.json()
